File: week-4/reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head
        begin = dummy
        i = 0
        while head:
            i += 1
            if i%k == 0:
                begin = self.reverse(begin, head.next)
                head = begin.next
            else:
                head = head.next
        
        return dummy.next
    
    def reverse(self, begin, end):
        prev = begin
        cur = begin.next
        first = cur
        
        while cur != end:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        
        begin.next = prev #list is reversed so prev is first node of slice
        first.next = cur # cur has moved to first node of next slice
        return first

File: week-4/reverse-nodes-in-k-group/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    let dummy = new ListNode(-1);
    dummy.next = head;
    let begin = dummy;
    let i = 0;
    while (head) {
        ++i;
        if (i%k === 0) {
            begin = reverse(begin, head.next);
            head = begin.next;
        } else
            head = head.next;
    }
    
    return dummy.next;
};

const reverse = (begin, end) => {
    let cur = begin.next;
    let prev = begin;
    let first = cur;
    
    while (cur !== end) {
        let temp = cur.next;
        cur.next = prev;
        prev = cur;
        cur = temp;
    }
    
    begin.next = prev;
    first.next = cur;
    return first;
};
